domain/robstoll-lib/atrium-domain-robstoll-lib-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/filesystem/symbolicLinkResolving.kt

```diff
@@ -0,0 +1,101 @@
+package ch.tutteli.atrium.domain.robstoll.lib.creating.filesystem
+
+import ch.tutteli.atrium.assertions.Assertion
+import ch.tutteli.atrium.assertions.AssertionGroup
+import ch.tutteli.atrium.domain.builders.ExpectImpl
+import ch.tutteli.atrium.reporting.RawString
+import ch.tutteli.atrium.translations.DescriptionPathAssertion.FAILURE_DUE_TO_LINK_LOOP
+import ch.tutteli.atrium.translations.DescriptionPathAssertion.HINT_FOLLOWED_SYMBOLIC_LINK
+import java.io.IOException
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.*
+
+inline fun explainForResolvedLink(path: Path, block: (realPath: Path) -> Assertion): Assertion {
+    val hintList = LinkedList<Assertion>()
+    val realPath = addAllLevelResolvedSymlinkHints(
+        path.toAbsolutePath(),
+        hintList
+    )
+    val blockResult = block(realPath)
+    return if (hintList.isNotEmpty()) {
+        when (blockResult) {
+            is AssertionGroup -> hintList.addAll(blockResult.assertions)
+            else -> hintList.add(blockResult)
+        }
+        ExpectImpl.builder.explanatoryGroup.withDefaultType
+            .withAssertions(hintList)
+            .build()
+    } else {
+        blockResult
+    }
+}
+
+/**
+ * Resolves the provided [absolutePath] and returns the resolved target (if resolving is possible).
+ * Adds explanatory hints for all involved symbolic links to [hintList].
+ */
+fun addAllLevelResolvedSymlinkHints(
+    absolutePath: Path,
+    hintList: Deque<Assertion>
+) = addAllLevelResolvedSymlinkHints(absolutePath, hintList, LinkedList(), absolutePath)
+
+private fun addAllLevelResolvedSymlinkHints(
+    absolutePath: Path,
+    hintList: Deque<Assertion>,
```

Comment:
why not a Hashset?

Reply:
For `hintList`: We never call `contains` or the like on it, using a `HashSet` would not bring any advantage, would it? (It would need to be `LinkedHashSet` because the hints are ordered)
For `visitedList`: We call `indexOf` on it, so i would not expect any gain from using a set, either.

Reply:
Sorry, making reviews on mobile, I intended to target visitedList not hintList
I think Set is more suitable for two reasons. 1) you don't want duplicate paths in the visitedList (I guess) and Set re-states this intention, this is the main reason IMO. 2) Set.contains is faster than LinkedList.indexOf as indexOf needs to travers the whole LinkedList if it isn't in there, having O(n) complexity in contrast to O(1) - however, the list will be small and only be used in case of an error where performance isn't key anyway.

Reply:
The key part here is that we want to call `indexOf`, and not `contains`! This is because `indexOf` gives us the number of items we can drop from `hintList` (because we do not want “followed symbolic link…” hints for links that a part of circle)

Regarding duplicates: We actually add a duplicated: the start of a loop. But this can be solved differently.

So would you still prefer a `LinkedHashSet`, even though the complexity will remain O(n)?

Reply:
Sorry, I am probably too tired now, I watched out for a use of visitedIndex other than `!= -1` and did not see it